pkg/stanza/fileconsumer/reader.go

```diff
@@ -85,23 +93,92 @@ func (r *Reader) ReadToEnd(ctx context.Context) {
 		}
 
 		token, err := r.encoding.Decode(scanner.Bytes())
-		if err != nil {
+		switch {
+		case err != nil:
 			r.Errorw("decode: %w", zap.Error(err))
-		} else {
-			r.emit(ctx, r.fileAttributes, token)
+		case r.headerSettings != nil && !r.HeaderFinalized:
+			if !r.consumeHeaderLine(ctx, r.FileAttributes, token) {
+				// recreate the scanner with the log-line's split func.
+				// We do not use the updated offset from the scanner,
+				// as the log line we just read could be multiline, and would be
+				// split differently with the new splitter.
+				if _, err := r.file.Seek(r.Offset, 0); err != nil {
+					r.Errorw("Failed to seek post-header", zap.Error(err))
+					return
+				}
+
+				scanner = NewPositionalScanner(r, r.maxLogSize, r.Offset, r.curSplitFunc())
+			}
+		default:
+			r.emit(ctx, r.FileAttributes, token)
 		}
 
 		r.Offset = scanner.Pos()
 	}
 }
 
+func (r *Reader) curSplitFunc() bufio.SplitFunc {
+	if r.headerSettings != nil && !r.HeaderFinalized {
+		return r.headerSettings.splitFunc
+	}
+
+	return r.splitFunc
+}
+
+// consumeHeaderLine checks if the given token is a line of the header, and consumes it if it is.
+// The return value dictates whether the given line was a header line or not.
+// If false is returned, the full header can be assumed to be read.
+func (r *Reader) consumeHeaderLine(ctx context.Context, attrs *FileAttributes, token []byte) bool {
+	if !r.headerSettings.matchRegex.Match(token) {
+		// Finalize and cleanup the pipeline
```

Comment:
This is where I'd assume we recreate the scanner and update the split func.

Reply:
Any suggestions on how we should handle recreating the scanner here? Because it's a local variable, it isn't available on the reader, so we don't have access here. 

EmitFunc doesn't have it as a parameter, we could maybe modify it to take a pointer to PositionalScanner, so we can swap in a new one (this doesn't feel quite right to me). 

Otherwise, we need to create a new struct with the EmitFunc every loop that has the scanner that we could modify.

Reply:
Yeah, I think now that you're calling `processFunc` it's simple enough.